File: leetcode/no0020_Valid_Parentheses.js
// link: https://leetcode.com/problems/valid-parentheses/
/**
 * @param {string} s
 * @return {boolean}
 */
 var isValid = function (s) {
    let result = true;
    let checkedChars = [];
    for (let i = 0; i < s.length; i++) {
        let char = s[i];

        if (char.match(/\(|\{|\[/) != null) {
            checkedChars.push(char);
        } else if (char.match(/\)|\}|\]/) != null) {
            let symbol = checkedChars.pop();
            let isBraceketMatched = true;
            if (char == ")" && symbol != "(") {
                isBraceketMatched = false;
            } else if (char == "}" && symbol != "{") {
                isBraceketMatched = false;
            } else if (char == "]" && symbol != "[") {
                isBraceketMatched = false;
            }

            if (!isBraceketMatched) {
                result = false;
                break;
            }
        }
    }
    if (checkedChars.length > 0) {
        result = false;
    }
    return result;
};

File: leetcode/no0020_Valid_Parentheses.go
package main

func isValid(s string) bool {

	checkedChars := []string{}
	result := true

	for i := 0; i < len(s); i++ {
		char := string(s[i])
		if char == "(" || char == "[" || char == "{" {
			checkedChars = append(checkedChars, char)
		} else if char == ")" || char == "]" || char == "}" {

			lastIndex := len(checkedChars) - 1
			if lastIndex < 0 {
				result = false
				break
			}
			lastChar := checkedChars[lastIndex]
			checkedChars = checkedChars[:len(checkedChars)-1]
			if (char == ")" && lastChar != "(") ||
				(char == "]" && lastChar != "[") ||
				(char == "}" && lastChar != "{") {
				result = false
				break
			}
		}
	}

	if len(checkedChars) > 0 {
		result = false
	}
	return result
}

func main() {
	//Input: s = "()"
	isValid("()")

}
